### PR TITLE
Replicate subspace_mlp_demo

### PR DESCRIPTION
**Pull request by karalleyna**
## Changes 

- [x] Add the following code snippet to subspace_sampler so that we can ignore warmup phase :) 
```
    if nsteps_full > 0 and nsteps_sub > 0:
        # Find good control variate / starting point in subspace
        params_tree, params_sub, log_post_trace, subspace_fns = subspace_optimizer(
            subspace_key, loglikelihood, logprior, params_init_tree, data, batch_size,
            subspace_dim, nsteps_full, nsteps_sub, opt, pbar=pbar)
        print(1, params_sub)
        loglik_sub, logprior_sub, subspace_to_pytree_fn = subspace_fns
    else:
        params_sub = jax.random.normal(subspace_key, (subspace_dim,))
        print(2, params_sub)
        if projection_matrix is None:
            projection_matrix = generate_random_basis(subspace_key, params_init_tree, subspace_dim)

        subspace_fns = make_subspace_fns(loglikelihood, logprior, params_init_tree, projection_matrix=projection_matrix)
```

## Results

If warmup phase is not discarded, test accuracies are as follows:

```

Testing subpace 10
Running time: 14.59s
Test Accuracy : 0.9698999524116516

Testing subpace 210
Running time: 18.32s
Test Accuracy : 0.9699999690055847

Testing subpace 410
Running time: 21.89s
Test Accuracy : 0.9696999788284302

Testing subpace 610
Running time: 25.04s
Test Accuracy : 0.9697999954223633
```

### CODE REVIEW

File: scripts/sgmcmc_subspace_lib.py [hunk 1]
```diff
@@ -69,22 +71,31 @@ def subspace_optimizer(key, loglikelihood, logprior, params_init_tree, data, bat
...
+            subspace_key, loglikelihood, logprior, params_init_tree, data, batch_size,
+            subspace_dim, nsteps_full, nsteps_sub, opt, pbar=pbar)
+        loglik_sub, logprior_sub, subspace_to_pytree_fn = subspace_fns
+    else:
+        params_sub = jax.random.normal(subspace_key, (subspace_dim,))
+        if projection_matrix is None:
+            projection_matrix = generate_random_basis(subspace_key, params_init_tree, subspace_dim)
+
+        subspace_fns = make_subspace_fns(loglikelihood, logprior, params_init_tree, projection_matrix=projection_matrix)
+
```
Comment (murphyk) [inline scripts/sgmcmc_subspace_lib.py:90]: Add
```
loglik_sub, logprior_sub, subspace_to_pytree_fn = subspace_fns
```
afterwards, otherwise undefined?

File: scripts/sgmcmc_subspace_lib.py [hunk 2]
```diff
@@ -69,22 +71,31 @@ def subspace_optimizer(key, loglikelihood, logprior, params_init_tree, data, bat
...
-    opt_key, sample_key = split(key)
-    params_tree, params_sub, log_post_trace, subspace_fns = subspace_optimizer(
-        opt_key, loglikelihood, logprior, params_init_tree, data, batch_size,
-        subspace_dim, nsteps_full, nsteps_sub, opt, pbar=pbar)
-    loglik_sub, logprior_sub, subspace_to_pytree_fn = subspace_fns
+                     subspace_dim, nsamples, opt=optax.adam(learning_rate=0.1),
+                     nsteps_full=0, nsteps_sub=0, projection_matrix=None, use_cv=True, pbar=True):
+    subspace_key, sample_key = split(key)
+
+    if nsteps_full > 0 and nsteps_sub > 0:
```
Comment (murphyk) [inline scripts/sgmcmc_subspace_lib.py:78]: I think this should be `if nsteps_full > 0 OR nsteps_sub > 0`

**Review by murphyk — CHANGES_REQUESTED** [review]

File: scripts/sgmcmc_subspace_lib.py (hunk 2 above)
Comment (karalleyna) [reply scripts/sgmcmc_subspace_lib.py]: Done

File: scripts/sgmcmc_subspace_lib.py (hunk 1 above)
Comment (karalleyna) [reply scripts/sgmcmc_subspace_lib.py:89]: Done :)
